frontend/main.js

```diff
@@ -175,6 +264,68 @@ async function startApp () {
     }
   })
 
+  sbp('okTurtles.events/on', EVENT_HANDLED, async (contractID, message) => {
+    // TODO: WRITE THIS MORE EFFICIENTLY SO THAT ONLY THE RELEVANT PARTS ARE
+    // COPIED INSTEAD OF THE ENTIRE CHELONIA STATE
+    const cheloniaState = await sbp('chelonia/rootState')
+    const state = cheloniaState[contractID]
+    const contractState = cheloniaState.contracts[contractID]
+    const vuexState = sbp('state/vuex/state')
+    if (contractState) {
+      if (!vuexState.contracts) {
+        Vue.set(vuexState, 'contracts', Object.create(null))
+      }
+      Vue.set(vuexState.contracts, contractID, cloneDeep(contractState))
+    } else if (vuexState.contracts) {
+      Vue.delete(vuexState.contracts, contractState)
+    }
+    if (state) {
+      Vue.set(vuexState, contractID, cloneDeep(state))
+    } else {
+      Vue.delete(vuexState, contractID)
+    }
+    sbp('okTurtles.events/emit', SHELTER_EVENT_HANDLED, contractID, message)
+  })
+
+  sbp('okTurtles.events/on', CONTRACTS_MODIFIED, async (subscriptionSet) => {
+    // TODO: WRITE THIS MORE EFFICIENTLY SO THAT ONLY THE RELEVANT PARTS ARE
+    // COPIED INSTEAD OF THE ENTIRE CHELONIA STATE
+    const cheloniaState = await sbp('chelonia/rootState')
+    const vuexState = sbp('state/vuex/state')
+
+    if (subscriptionSet.includes(vuexState.pendingGroupId)) {
+      sbp('state/vuex/commit', 'setCurrentGroupId', { contractID: vuexState.pendingGroupId })
+    }
+
+    if (!vuexState.contracts) {
+      Vue.set(vuexState, 'contracts', Object.create(null))
+    }
+
+    const oldContracts = Object.keys(vuexState.contracts)
+    const oldContractsToRemove = oldContracts.filter(x => !subscriptionSet.includes(x))
+    const newContracts = subscriptionSet.filter(x => !oldContracts.includes(x))
+
+    oldContractsToRemove.forEach(x => {
+      Vue.delete(vuexState.contracts, x)
+      Vue.delete(vuexState, x)
+    })
+    newContracts.forEach(x => {
+      const state = cheloniaState[x]
+      const contractState = cheloniaState.contracts[x]
+      if (contractState) {
+        Vue.set(vuexState.contracts, x, cloneDeep(contractState))
+      }
+      if (state) {
+        Vue.set(vuexState, x, cloneDeep(state))
+      }
+    })
+  })
```

Comment:
Please see if this code can be rewritten to be more generic (to support redux etc) and moved into some chelonia namespace that's imported. We cannot expect developers to write this.

Reply:
I think it can be made more generic, at least as generic as chelonia is with `reactiveSet`, etc. It already is pretty generic, it's just needing a way to call / include it. The parts to move are these, the saveCheloniaState part and maybe some of the login handler.

Reply:
Moved